### ISSUE TITLE
Directions and Priorities for 0.12.0

### ISSUE THREAD

**Issue by andy-maier** [pywbem/pywbem#771]
In PR #330, we stated that we want 0.12.0 to be the last release supporting Python 2.6. In order to also make this a rather quick release, the following list of items attempts to do only what can be done easily. The larger things are moved out to a 1.0.0 version.

Karl and Andy agreed on the following list, in our call on 2017-10-04:

Items for 0.12.0:

* Review current experimental functionality for finalization:
  - Iter support (`WBEMConnection.Iter...()` methods & iter-related shorthands in `wbemcli`). YES. (issue #875). Note issue #782 which needs to be fixed before we can/should finalize.
    Status: DONE (PR #878, merged)
  - Statistics support (classes `OperationStatistic` and `Statistics`, `enable_stats` init arg on `WBEMConnection`, `stats_enabled`, `statistics`, `last_operation_time` and `last_server_response_time` properties on `WBEMConnection`). YES. (issue #876)
    Status: DONE (PR #877, merged)
  - Operation recorder support (`operation_recorder` property on `WBEMConnection` [marked both experimental and deprecated], `operation_recorder_enabled` property on `WBEMConnection`).
    Status: NOT YET.
  - Logging support.
    Status: NOT YET.
  - WBEM server API. Note that the change log for 0.10.0 states: "Declared the WBEM indications API and the WBEM server API to be final. These APIs had been introduced in v0.9.0 as experimental." However, the description of WBEM server API still marks it as experimental. Do we want to consider this a documentation bug? YES. (issue #873)
    Status: DONE (PR #880, merged)
  - `irecv` namespace (still mentioned in DEVELOP.md) YES (just wording to be fixed). (issue #874)
    Status: DONE (PR #879, merged)
*  Issues with use_pull_operations attribute (issue #782) - needed for finalizing the Iter support
  Status: DONE (PR #881, merged)
* Unicode / binary string inconsistencies between py3 and py2 (issue #746)
  Status: DONE - no issue
* Fix that compiler requires a destination even for complete compile and dryrun (issue #424)
  Status: DONE - moved to after 0.12
* Remove print output from test_statistics (issue #779)
  Status: DONE (PR #803, merged)
* cim_http validates https but not http (issue #457)
  Status: DONE (won't fix)
* Handle the XML prefix string on http output (issue #418)
  Status: DONE -> Moved beyond 0.12
* OpenQueryInstances has incorrectly named query parameters (issue #781)
  Status: DONE - no issue
* Fix mandatory bug: Indication listener reuse fails in python 3 (issue #452 -> now handled in #1090)
  Status: DONE (PR #1097, merged)
* Complete documentation review and fixing (create individual PRs) (issue #1067)
  Status: DONE (Multiple PRs as stated in the issue)
* Reduce Pylint findings.
  Status: Addressed some more in PRs: #811, #820-831, #934, #1001, #1075, #1102
* Fix some more (optional?) bugs: [Open issues marked as bugs](https://github.com/pywbem/pywbem/issues?q=is%3Aissue+is%3Aopen+label%3A%22type%3A+bug%22)
  Status: DONE (All bugs targeted for 0.12 are resolved now)
* Provide updates for [Open issues marked as update needed](https://github.com/pywbem/pywbem/issues?utf8=%E2%9C%93&q=is%3Aopen%20is%3Aissue%20label%3A%22update%20needed%22%20).
  Status: DONE (No more update needed issues targeted for 0.12)

Items for 1.0.0:

* Add new class-level association operations (issue #349).
* Extensions to server and indications:
  - Jobs
  - Support WBEM Server profile 
  - TBD: more ideas / items are needed
* Return to a single SSL environment for both py2 and py3 (issue #496).
  - Possibly with an option for users to still use M2Crypto on py2
* Improve efficiency/speed. Some ideas are:
  - Move away from minidom in pretty-printing (issue #579).
  - Move from SAX to lxml (downside: another package with compiled code) (issue #300).
  - Move from `httplib` to `requests` (issue #280).
  - Consider moving away from the tupleparser as an interim object.
  - Generate tests to compare Pegasus client with pywbem client in order to determine how big the issue is w.r.t. efficiency and speed. Can be measured with the new statistics support in 0.11.0.
* Bring pywbem makefile/setup more into line with other repositories such as pywbemtools, zhmc (in long term we should have about the same install/build infrastructure in all repositories in pywbem)
  - use requirements.txt files
  - change make check to execute just flake8 (no longer pylint), and have pylint in a separate target
  - use of pbr for automatic versioning (conflict with py 2.6)
* Migrate tests to use py.test:
  - for simple tests like cim_obj and cim_types (easy to start with, better coverage possible)
  - for testclient (high benefit)
  - then all the rest
* Improve test coverage. Currently we are at 71% for Python 2.7. [Open issues marked as test enhancements](https://github.com/pywbem/pywbem/issues?utf8=%E2%9C%93&q=is%3Aopen%20is%3Aissue%20label%3A%22area%3A%20test%22%20label%3A%22type%3A%20enhancement%22%20)
* Remove the `**extra` argument from the `WBEMConnection` operation methods. It has no use except if someone implemented private extension in a server, and is error-prone in case a parameter name is mistyped (happened a few times!).
